src/routes/transactions/mappers/common/twap-order.mapper.ts

```diff
@@ -195,37 +193,114 @@ export class TwapOrderMapper {
-      if (i === orders.length - 1) {
-        return order.status;
+    for (const part of args.partsToFetch) {
+      const partFullAppData = await this.swapsRepository.getFullAppData(
+        args.chainId,
+        part.appData,
+      );
+
+      if (!this.swapAppsHelper.isAppAllowed(partFullAppData)) {
+        throw new Error(
+          `Unsupported App: ${partFullAppData.fullAppData?.appCode}`,
+        );
       }
 
-      // If the part is complete, continue to the next part
-      if (completeStatuses.includes(order.status)) {
-        continue;
+      const orderUid = this.gpv2OrderHelper.computeOrderUid({
+        chainId: args.chainId,
+        owner: args.safeAddress,
+        order: part,
+      });
+      const order = await this.swapsRepository
+        .getOrder(args.chainId, orderUid)
+        .catch(() => {
+          this.loggingService.warn(
+            `Error getting orderUid ${orderUid} from SwapsRepository`,
+          );
+        });
+
+      if (!order || order.kind == OrderKind.Unknown) {
+        // Without every order it's not possible to determine executed amounts/fees
+        return null;
       }
 
-      return order.status;
+      if (!this.swapAppsHelper.isAppAllowed(order)) {
+        throw new Error(`Unsupported App: ${order.fullAppData?.appCode}`);
+      }
+
+      orders.push(order as KnownOrder);
+    }
+
+    return orders;
+  }
+
+  private async getOrderStatus(args: {
+    chainId: string;
+    safeAddress: `0x${string}`;
+    twapParts: Array<GPv2OrderParameters>;
+    partOrders: Array<KnownOrder> | null;
+    activeOrderUid: `0x${string}` | null;
+    executionDate: Date | null;
+  }): Promise<OrderStatus> {
+    if (!args.executionDate) {
+      return OrderStatus.PreSignaturePending;
     }
 
-    return OrderStatus.Unknown;
+    // Note: TWAPs can never "expire" in their entirety, only parts can but
+    // this does not affect TWAP status
+
+    const finalOrderUid = this.gpv2OrderHelper.computeOrderUid({
+      chainId: args.chainId,
+      owner: args.safeAddress,
+      order: args.twapParts.slice(-1)[0], // Last TWAP part
+    });
+
+    // If active order is final one, it's fulfilled as it exists
+    if (args.activeOrderUid === finalOrderUid) {
+      return OrderStatus.Fulfilled;
+    }
+
+    // Check active or final order part depending on whether the TWAP is active
+    const orderUidToCheck = args.activeOrderUid ?? finalOrderUid;
+
+    try {
+      // If we already fetched the order, we don't need to again
+      const orderAlreadyFetched = !!args.partOrders?.some(
+        (order) => order.uid === orderUidToCheck,
+      );
+      if (!orderAlreadyFetched) {
+        await this.swapsRepository.getOrder(args.chainId, orderUidToCheck);
+      }
+
+      // If the order exists, the TWAP is either open or fulfilled depending
+      // on whether the order is active
+      return args.activeOrderUid ? OrderStatus.Open : OrderStatus.Fulfilled;
```

Comment:
I think we could remove the above check (where I commented) and instead do something like this here:
```suggestion
      return args.activeOrderUid ? args.activeOrderUid === finalOrderUid ? OrderStatus.Fulfilled : OrderStatus.Open : OrderStatus.Fulfilled;
```

Reply:
Adjusted in f20a158137e033f1620ec10484a6501a82057e7f.